### PR TITLE
Make sure DDPM and `diffusers` can be used without Transformers

### PR DESCRIPTION
**Pull request by sayakpaul**
Fixes https://github.com/huggingface/diffusers/issues/5661. 

I think we need to make a patch release for this one. 

While reviewing, could y'all please go over [this comment](https://github.com/huggingface/diffusers/issues/5661#issuecomment-1797391247) too and verify things on your ends?

### CODE REVIEW

**Review by DN6 — APPROVED** [review]

**Comment by DN6** [comment]
@sayakpaul `pip install -e "[.test]" [installs transformers](https://github.com/huggingface/diffusers/blob/9bafef34bde99a6184c4d4c5af8bd434244a5ed9/setup.py#L221), which is why `tests/others/test_dependencies.py` didn't catch it.

**Comment by sayakpaul** [comment]
But I am doing `pip install -e .` here:

https://github.com/huggingface/diffusers/blob/9bafef34bde99a6184c4d4c5af8bd434244a5ed9/.github/workflows/pr_dependency_test.yml#L27

**Comment by DN6** [comment]
Hmm that should have been caught then. Checking the container.

**Comment by DN6** [comment]
Do we even need to import `PretrainedModel` and `PretrainedTokenizer` if it's just a string that's being used for type hints?

**Comment by sayakpaul** [comment]
> Do we even need to import `PretrainedModel` and `PretrainedTokenizer` if it's just a string that's being used for type hints?

I think it's for type hints.

**Comment by DN6** [comment]
Hmm so the tests pass because torch isn't installed either, so just dummy objects are being created 
https://github.com/huggingface/diffusers/actions/runs/6780411004/job/18429004601?pr=5671#step:5:20

**Comment by sayakpaul** [comment]
Wonder if there is a more robust method to test the dependencies here. @DN6 what would you suggest?

**Comment by DN6** [comment]
I think there are a couple of issues 

Models are all imported if torch is available otherwise we get dummies, but the Mixins can introduce additional dependencies on transformers or other soft dependencies. (Technically torch is optional too) 

So we should be careful about how we add Mixins into model classes if they are only supposed to depend on a single dependency (torch or jax).

Adds dedicated workflows for the soft dependencies.

File: src/diffusers/models/lora.py [hunk 1]
```diff
@@ -18,14 +18,15 @@
 import torch.nn.functional as F
 from torch import nn
 
-from ..loaders import PatchedLoraProjection, text_encoder_attn_modules, text_encoder_mlp_modules
```
Comment (patrickvonplaten) [inline src/diffusers/models/lora.py:21]: Why do we need to do this change? Don't think it's necessary no?

File: src/diffusers/loaders.py [hunk 2]
```diff
@@ -2390,7 +2390,7 @@ def unfuse_text_encoder_lora(text_encoder):
     def set_adapters_for_text_encoder(
```
Comment (patrickvonplaten) [inline src/diffusers/loaders.py:2390]: Nice! Going forward, let's make sure we always use strings for Transformers type hints

File: src/diffusers/pipelines/latent_consistency_models/__init__.py [hunk 3]
```diff
@@ -1,19 +1,40 @@
 from typing import TYPE_CHECKING
 
 from ...utils import (
+    DIFFUSERS_SLOW_IMPORT,
```
Comment (patrickvonplaten) [inline src/diffusers/pipelines/latent_consistency_models/__init__.py:4]: Very nice clean-up!

File: src/diffusers/pipelines/pixart_alpha/__init__.py [hunk 4]
```diff
@@ -1 +1,48 @@
-from .pipeline_pixart_alpha import PixArtAlphaPipeline
+from typing import TYPE_CHECKING
+
```
Comment (patrickvonplaten) [inline src/diffusers/pipelines/pixart_alpha/__init__.py:2]: Very nice clean-up!

**Comment by patrickvonplaten** [comment]
PR is good to go for me, just wondering whether we really need to do this change: https://github.com/huggingface/diffusers/pull/5668#discussion_r1385126853

**Pull request merged**